Add Ignore Author Server button to post actions

MrPetovan opened the pull request:
Closes https://github.com/friendica/friendica/issues/13286

This features hides top-level posts from the muted servers in the Global Community page, the network page and returns a Not Found error in the display page when trying to open top-level posts from servers that were muted.

MrPetovan committed:
[Database 1526] Add new user-gserver table

- Add Entity, Collection, Factory and Repository classes for it

MrPetovan committed:
[frio] Add Mute Author Server button to post actions

MrPetovan committed:
Create DDD classes for gserver class

MrPetovan committed:
Add user settings page to manage remote server settings

MrPetovan committed:
Updated main translation file after adding several strings

MrPetovan committed:
[Database 1527] Add required gsid fields to post-user-view, post-thread-user-view, post-view and  post-thread-view

MrPetovan committed:
Fix DisplayNotFound page

- Add another condition earlier in Module\Item\Display
- Separate reasons in individual translation strings

MrPetovan commented on src/Module/Conversation/Community.php:
```diff
@@ -356,7 +356,7 @@ private static function selectItems($min_id, $max_id, $item_id, $itemspage)
 		}
 
-		$r = Post::selectThreadForUser(0, ['uri-id', 'commented', 'author-link'], $condition, $params);
+		$r = Post::selectThreadForUser(DI::userSession()->getLocalUserId() ?: 0, ['uri-id', 'commented', 'author-link'], $condition, $params);
```
@annando This change allows for remote server local user settings to be taken into account in the Global Community page. Can you see any drawbacks of doing this?

annando replied:
One thing that I see here is that this only can cover posts that are started by a blocked server, excluding replies from contacts on blocked servers.

MrPetovan replied:
How would I go about that? I couldn't find an obvious place where replies are fetched in the Community module.

annando replied:
See here: https://github.com/friendica/friendica/blob/develop/src/Content/Conversation.php#L955-L992C4

MrPetovan replied:
Thanks, the behavior now matches the description.

MrPetovan commented:
I forgot about notifications coming from muted servers.

MrPetovan commented:
Notifications have been taken care of.

MrPetovan committed:
Remove DI dependency from Module\Contact\Profile

MrPetovan committed:
[vier] Add mute author server link in photo menu

MrPetovan commented:
I've added vier support to this PR.

annando commented on database.sql:
```diff
@@ -2707,7 +2728,9 @@ CREATE VIEW `network-thread-view` AS SELECT
 			AND (`author`.`blocked` IS NULL OR NOT `author`.`blocked`)
-			AND (`owner`.`blocked` IS NULL OR NOT `owner`.`blocked`);
+			AND (`owner`.`blocked` IS NULL OR NOT `owner`.`blocked`)
+			AND NOT `authorcontact`.`gsid` IN (SELECT `gsid` FROM `user-gserver` WHERE `uid` = `post-thread-user`.`uid` AND `muted`)
```
Can you change this to an `AND NOT EXISTS(SELECT`? This seems to be better concerning performance.

MrPetovan replied:
Done.

annando commented on static/dbstructure.config.php:
```diff
@@ -159,6 +159,17 @@
+		"fields" => [
+			"uid" => ["type" => "mediumint unsigned", "not null" => "1", "default" => "0", "foreign" => ["user" => "uid"], "comment" => "Owner User id"],
+			"gsid" => ["type" => "int unsigned", "not null" => "1", "default" => "0", "foreign" => ["gserver" => "id"], "comment" => "Gserver id"],
+			"muted" => ["type" => "boolean", "not null" => "1", "default" => "0", "comment" => "server accounts are muted for the user"],
```
With the users we have got `blocked`, `ignored` and `collapsed`. Shouldn't we keep that theme here as well?

MrPetovan replied:
What theme are you referring to?

annando replied:
I guess "schema" is the better word.

MrPetovan replied:
Very well, what schema are you referring to? What do you think I should rename this field to?

annando replied:
I mean that we should use the same words that we use for contacts.

MrPetovan replied:
Got it. None of the words we use for contacts matches this feature. Collapse and Ignore don’t go that far that I know of, and Block goes further. So I thought it warranted a new term to avoid feature confusion.

annando replied:
I think that this matches really well the definition of "ignoring", since we are only blocking the display of incoming posts from the given domains.

MrPetovan replied:
It’s close, but I seem to remember we still show replies from ignored authors, while we don’t from authors of muted remote servers.

annando replied:
AFAIK we don't show ignored authors on the network page, but only on the display page, don't we?

MrPetovan replied:
Possibly, which makes it one more place we show ignored authors than muted server authors.

annando replied:
So with this PR we have a "mute" feature that mostly acts like the "ignore" feature for contacts and in the future will be called by the Mastodon API endpoints for blocked servers. And since we already have got the functionality in the conversation class to handle blocked, ignored and collapsed contacts, in the future we could easily extent this to the servers as well, so that we would have blocked, ignored, collapsed and muted servers.

I think that this is confusing for the users to understand. That's why I think that for the servers we should stick with the terminology for the contacts.

MrPetovan replied:
Why not, but then we should align the `ignored` contact behavior to the now-`ignored` server behavior, meaning that posts from an ignored contact should never show in any circumstance, but can still be received as it only affects the display.

Now that I'm writing this, I don't think the check I added in `Protocol\ActivityPub::isAcceptedRequester` is warranted for now-`ignored` domains.

annando commented on src/Model/Post.php:
```diff
@@ -457,8 +457,11 @@ private static function selectViewForUser(string $view, int $uid, array $selecte
 			AND NOT `author-id` IN (SELECT `cid` FROM `user-contact` WHERE `uid` = ? AND `ignored` AND `cid` = `author-id`)
-			AND NOT `owner-id` IN (SELECT `cid` FROM `user-contact` WHERE `uid` = ? AND `ignored` AND `cid` = `owner-id`)",
-				0, Contact::SHARING, Contact::FRIEND, 0, $uid, $uid, $uid, $uid, $uid]);
+			AND NOT `owner-id` IN (SELECT `cid` FROM `user-contact` WHERE `uid` = ? AND `ignored` AND `cid` = `owner-id`)
```
At other places we used `EXISTS` - which improved the performance. Possibly we should use that format everywhere? https://github.com/friendica/friendica/blob/develop/src/Worker/ExpirePosts.php#L182-L197

MrPetovan replied:
Does it mean changing all other sub-queries here and in the other queries I changed?

annando replied:
I'm currently unsure. Can you do some tests with `explain`?

MrPetovan replied:
With some reluctance, yes. Manually replacing the placeholders with the actual values is painstaking in large queries like these.

annando replied:
If not, then it is okay as well. I guess I have a larger database at hand.

MrPetovan replied:
I’ve now changed the whole condition, and it still works. No idea about the performance change.

annando commented:
Update: I'm still in the process of reviewing the code, but 38 files are 38 files ...

MrPetovan commented:
As usual, I separated concerns between commits, I recommend going through each one separately if 38 files is too much at once.

annando commented:
Can you add a server check here? https://github.com/friendica/friendica/blob/develop/src/Protocol/ActivityPub.php#L313

MrPetovan committed:
Implement ignored server block in conversations

- Add server ignore status in contact profile page
- Add new reason in DisplayNotFound exception page

MrPetovan commented:
> Can you add a server check here? https://github.com/friendica/friendica/blob/develop/src/Protocol/ActivityPub.php#L313

Done.

MrPetovan commented:
After verifying the user ignore feature doesn't have the caveat I thought it had, I renamed the server feature to ignore as well.

annando commented on src/Database/Database.php:
```diff
@@ -849,29 +849,11 @@ public function exists(string $table, array $condition): bool
-		$this->close($stmt);
-
-		return $retval;
+		return $this->count($table, $condition) > 0;
```
A `count` is significantly slower than just having a look if any database entry exists.

MrPetovan replied:
Unfortunately there's no "just having a look if any database entry exists" function in MySQL/MariaDB (I looked for one), so either way (Statement row number or COUNT) can be defended. As I understand it, `LIMIT` is mainly useful to reduce the amount of data being transferred from the database to the client, but it can slow down queries. This is inconvenient here as we don't need the query data at all.

I prefer the COUNT version as it is simpler both in code and in concept, you're "just" asking the database itself the number of records in the result set produced by the provided condition.

annando replied:
`count` will be deadly for large tables, since the system has to scan through the whole database to find each entry. The query with the limit stops at the first occurrence, which will be much faster. See this example:
```
]> select count(*) from `post-user-view` where body like "%Test%";
+----------+
| count(*) |
+----------+
|   265487 |
+----------+
1 row in set (27 min 48,265 sec)

> select id from `post-user-view` where body like "%Test%" limit 1;
+----------+
| id       |
+----------+
| 20406627 |
+----------+
1 row in set (0,004 sec)
```

MrPetovan replied:
Oof, can’t argue with these numbers. I’ll revert my changes.

MrPetovan replied:
Done.

MrPetovan commented:
@nupplaphil Codecov seems misconfigured at the moment:
```
+ codecov -R '.' -Z -f 'clover.xml'
[[2](https://ci.friendi.ca/repos/2/pipeline/59/46#L2)02[3](https://ci.friendi.ca/repos/2/pipeline/59/46#L3)-08-17T10:57:[4](https://ci.friendi.ca/repos/2/pipeline/59/46#L4)2.380Z] ['info'] 
     _____          _
    / ____|        | |
   | |     ___   __| | ___  ___ _____   __
   | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
   | |___| (_) | (_| |  __/ (_| (_) \ V /
    \_____\___/ \__,_|\___|\___\___/ \_/

  Codecov report uploader 0.4.1
[2023-08-17T10:[5](https://ci.friendi.ca/repos/2/pipeline/59/46#L5)7:42.381Z] ['info'] => Project root located at: .
[2023-08-17T10:57:42.383Z] ['info'] ->  Token found by environment variables
[2023-08-17T10:57:42.483Z] ['info'] Searching for coverage files...
[2023-08-17T10:57:42.[6](https://ci.friendi.ca/repos/2/pipeline/59/46#L6)01Z] ['info'] => Found 1 possible coverage files:
  clover.xml
[2023-08-1[7](https://ci.friendi.ca/repos/2/pipeline/59/46#L7)T10:57:42.601Z] ['info'] Processing clover.xml...
[2023-0[8](https://ci.friendi.ca/repos/2/pipeline/59/46#L8)-17T10:57:42.65[9](https://ci.friendi.ca/repos/2/pipeline/59/46#L9)Z] ['info'] Detected Woodpecker CI as the CI provider.
[2023-08-17T[10](https://ci.friendi.ca/repos/2/pipeline/59/46#L10):57:42.660Z] ['info'] Pinging Codecov: https://codecov.io/upload/v4?package=uploader-0.4.1&token=*******&branch=task%2Fuser-server-block&build=59&build_url=https%3A%2F%2Fgithub.com%2Ffriendica%2Ffriendica%2Fpull%2F[13](https://ci.friendi.ca/repos/2/pipeline/59/46#L13)288&commit=0eb16eccd40d8aff2217ad998b320c97825460[14](https://ci.friendi.ca/repos/2/pipeline/59/46#L14)&tag=&pr=13288&job=0&service=woodpecker&slug=friendica%2Ffriendica&name=&flags=&parent=
[2023-08-[17](https://ci.friendi.ca/repos/2/pipeline/59/46#L17)T10:57:42.856Z] ['error'] There was an error running the uploader: Error uploading to https://codecov.io: Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Could not find a repository associated with upload token ********', code='not_found')}
```

nupplaphil commented:
Yeah I know, but unfortunately I need a new token from/for codecov, which only works with the gihtub friendica account. @tobiasd will create one when he has some time for it

MrPetovan commented:
> > Can you add a server check here? https://github.com/friendica/friendica/blob/develop/src/Protocol/ActivityPub.php#L313
> 
> Done.

I finally decided to go against it. Ignoring only hides content, it shouldn't prevent any communication at the protocol level. The `@TODO` hint mentions "server block" and this isn't it (yet).

annando commented:
The check there is not comparable with a server block. We are not blocking communication with the server there. We are just preventing to transmit any personal data that isn't needed. So this is softer than a block.

MrPetovan commented:
"Ignore" is the equivalent of a shadow-block. This means that the Friendica user can't see anything from the ignored account, but the ignored account doesn't get any technical hint that they have been moderated by the Friendica user. This means "Ignore" shouldn't act at protocol level no matter how "soft" it is.

annando commented:
To keep the order of the PRs, I would like to merge it, after https://github.com/friendica/friendica/pull/13339 is merged. The only problem there is the failing database check.

annando commented:
As expected, there are now some conflicts.

MrPetovan commented:
Yup, running xgettext at the moment, it will be ready soon.
